test/integration/mqtt_paho_binding/mqtt_test.go

```diff
@@ -2,139 +2,130 @@ package mqtt_paho_binding
-				eventOut := MustToEvent(t, ctx, out)
-				AssertEventEquals(t, eventIn, ConvertEventExtensionsToString(t, eventOut))
-			})
-	})
-}
-
-func TestSendBinaryMessageToBinary(t *testing.T) {
-	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
-	s, r, c, err := getSenderReceiver(ctx, cancel)
-	require.NoError(t, err)
-	defer c()
-
-	EachEvent(t, Events(), func(t *testing.T, eventIn event.Event) {
-		eventIn = ConvertEventExtensionsToString(t, eventIn)
-		in := MustCreateMockBinaryMessage(eventIn)
+func TestEncodingBinding(t *testing.T) {
+	testCases := []struct {
+		name             string
+		inEncoding       binding.Encoding
+		inEventEncoder   func(tt *testing.T, inEvent event.Event) binding.Message
+		expectedEncoding binding.Encoding
+	}{
+		{
+			name:       "Structured message with Structured encoding",
+			inEncoding: binding.EncodingStructured,
+			inEventEncoder: func(tt *testing.T, inEvent event.Event) binding.Message {
+				return MustCreateMockStructuredMessage(tt, inEvent)
+			},
+			expectedEncoding: binding.EncodingStructured,
+		},
+		{
+			name:       "Binary message with Binary encoding",
+			inEncoding: binding.EncodingBinary,
+			inEventEncoder: func(tt *testing.T, inEvent event.Event) binding.Message {
+				return MustCreateMockBinaryMessage(inEvent)
+			},
+			expectedEncoding: binding.EncodingBinary,
+		},
+	}
 
-		sendReceive(t, binding.WithPreferredEventEncoding(ctx, binding.EncodingBinary), s, r, in,
-			func(out binding.Message) {
-				assert.Equal(t, binding.EncodingBinary, out.ReadEncoding())
-				eventOut := MustToEvent(t, ctx, out)
-				AssertEventEquals(t, eventIn, ConvertEventExtensionsToString(t, eventOut))
+	for _, tt := range testCases {
+		t.Run(tt.name, func(t *testing.T) {
+			EachEvent(t, Events(), func(t *testing.T, inEvent event.Event) {
+				ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+				defer cancel()
+				topicName := "test-ce-client-" + uuid.New().String()
+
+				// start a cloudevents receiver client go to receive the event
+				receiveChan := make(chan receiveMessage)
+				defer close(receiveChan)
+				startReceiver(ctx, topicName, receiveChan)
+
+				// start a cloudevents sender client go to send the event
+				sender, err := getProtocol(ctx, topicName)
+				require.NoError(t, err)
+				defer sender.Close(ctx)
+
+				inEvent = ConvertEventExtensionsToString(t, inEvent)
+				inMessage := tt.inEventEncoder(t, inEvent)
+				ctx = binding.WithPreferredEventEncoding(ctx, tt.inEncoding)
+
+				for {
+					select {
+					case <-ctx.Done():
+						require.Fail(t, "timeout waiting for event")
+						return
+					case receiveMsg := <-receiveChan:
+						require.NoError(t, receiveMsg.err)
+						outMessage := receiveMsg.msg
+						assert.Equal(t, tt.expectedEncoding, outMessage.ReadEncoding())
+						outEvent := MustToEvent(t, ctx, outMessage)
+						AssertEventEquals(t, inEvent, ConvertEventExtensionsToString(t, outEvent))
+						return
+					default:
```

Comment:
nit: just be aware that this is a hot loop now, ideally should also be a ticker, right?

Reply:
@embano1 Yes, It's a hot loop! 
Since the `default` case causes a bit of confusion here, I turned it into an explicit ticker.